core/node/eth_sender/src/aggregator.rs

```diff
@@ -275,6 +311,84 @@ impl Aggregator {
         self.priority_merkle_tree.as_mut().unwrap()
     }
 
+    async fn get_precommit_operation(
+        &mut self,
+        storage: &mut Connection<'_, Core>,
+        precommit_params: PrecommitParams,
+    ) -> Result<Option<L2BlockAggregatedOperation>, EthSenderError> {
+        let last_committed_l1_batch = storage
+            .blocks_dal()
+            .get_number_of_last_l1_batch_committed_on_eth()
+            .await
+            .unwrap();
+
+        let last_sealed_l1_batch = storage
+            .blocks_dal()
+            .get_sealed_l1_batch_number()
+            .await
+            .unwrap();
+
+        // If we have some sealed, but not committed L1 batches, we need to precommit all txs from this batch,
+        // before sending the commit tx. If all batches were committed and we have only open batch,
+        // we have to start committing the txs from it (last_sealed_l1_batch == last_committed_l1_batch).
+        // If we have no committed  batches, we need to precommit the txs from the first batch.
+        let (actual_batch_number, desired_batch_number_for_db) =
+            match (last_sealed_l1_batch, last_committed_l1_batch) {
+                (None, _) => return Ok(None),
+                (Some(last_sealed_l1_batch), Some(last_committed_l1_batch))
+                    if last_sealed_l1_batch == last_committed_l1_batch =>
+                {
+                    (last_sealed_l1_batch + 1, None)
+                }
+                (Some(_), Some(committed_l1_batch_number)) => (
+                    committed_l1_batch_number + 1,
+                    Some(committed_l1_batch_number + 1),
+                ),
+                (Some(_), None) => (1.into(), Some(1.into())),
+            };
+
+        let txs = storage
+            .blocks_dal()
+            .get_ready_for_precommit_txs(desired_batch_number_for_db)
```

Comment:
When desired_batch_number_for_db is NULL there may be race condition, e.g. you have last_sealed_l1_batch == last_committed_l1_batch = 10, then pending L1 batch 11 is sealed, then you do get_ready_for_precommit_txs(None) and get results for L1 batch 12 which is now pending. I don't have immediate suggestions how to fix it

Reply:
@perekopskiy, i found a solution for this. 
Meanwhile, i noticed, that disabling precommits, doesn't properly work, will address it, but the main  code won't change